### PR TITLE
Streaming Backend for Service Health

### PR DESCRIPTION
**Pull request by banks**
This is the culmination of almost a year's worth of work in a feature branch (streaming-rpc).

That branch was so out of date due to two major refactors in the code in the interim (Namespaces and hclog) that it was impossible to just merge. Not only that but Namespaces also introduced tooling and conventions for protobuf in the agent which we need to convert streaming code to use too.

This mega PR is a painstaking re-factor of all the code in `streaming-rpc-optimizations` branch. It supersedes the refactoring work done in #7128 and #7143 making them obsolete.

## PR Approach

I realise this is a huge PR and does contain some new code. Since a straight forward merge was impossible I took the approach of porting all the logical components from the `streaming-rpc-optimizations` branch and getting them to work against master in isolation.

Each commit in the PR history is one logical component so hopefully makes it a little easier to digest. I'll also note which ones changed significantly below to aid review.

## Notable Changes

Most of the design of the streaming backend is unchanged from previously reviewed code in the feature branch or in #7128.

The biggest differences added here that have not been seen before are:

### Protobuf Code Gen

I moved the protobuf components of streaming into the `agentpb` package that was added as part of Namespaces work.

Also, I felt very uncomfortable about having proto defintions that needed to be kept manually in sync with `structs` package especially when all of the core data types will eventually end up mirrored and are changing often as features are developed.

So in c5f5a05 (modified slightly later in 995327cb5) I spent a _long_ time writing code generation that will generate conversion methods between structs and protobuf versions as well as generating tests that will catch and fail on any changes to one package that are not mirrored in the other. **These have already caught changes several times during development** that we wouldn't otherwise have noticed and so are proving valuable already.

### State Store Transactions

The biggest change here to the code base is in the `state` package in 7965767. While working on the state store event emitting from the branch and porting it, I found 6 or 7 bugs in the way events were being emitted for service registrations alone. I spent several days refactoring the code to make it correct and what I ended up with was so complicated it felt worryingly unmaintainable and brittle especially once we port this to all other blocking queries and types in the state store. I assumed at first that service health was just the worst case and others would be simpler however just trying to get the most basic events emitted for ACL proved that wasn't true.

The main issue is that we tend to have multiple code paths for modifying most record types in the state store. So multiple public methods tend to call into internal methods that do the actual mutation, but these are composed in different ways. The logical place to emit events is right at the point where we insert into memDB e.g. in `ensureNodeTxn` but at that point we don't have context for whether we are also updating services on that node or checks on those services etc.

So instead of manually threading context tracking code through the spaghetti calls inside state store for this, I move the problem down a level of abstraction.

I implemented a change to MemDB itself in https://github.com/hashicorp/go-memdb/pull/70 and https://github.com/hashicorp/go-memdb/pull/71 that allows MemDB to report a summary of all the table changes committed in each transaction.

**We now use `*txnWrapper` instead of `*memdb.Txn` everywhere** in the state store. The wrapper enforces a few things - that we know the index the commit is happening at, and that changes are always tracked for every write transaction. It also cause the transaction `Commit` to pass the actual final changes from MemDB (note, regardless of the code path they took through statestore) through a registered set of callbacks that all get a chance to inspect the changes and emit any events necessary.

This is way cleaner and more maintainable:
 - Whatever we change in the state store code, we can't miss a code path that updates some state.
 - Event emitting logic is nicely decoupled from transaction logic
 - Event emitting logic can all live in one, easily unit testable function rather than be threaded through state store transaction code.

### TODO After Merge

This is _pretty close_. I think it's mergeable as is (except for any review issues) with a few caveats before 1.8.0:

 - There are a bunch of TODOs around Namespace support. Especially things like separate protobuf structs in Enterprise vs OSS. I didn't want to solve for that as well as all the rest of this so I propose to merge this to OSS master as-is, and then work on the Namespace support on top.
 - We probably need a bunch more documentation eventually - the config and new telemetry are documented but we probably need a whole internals section about this eventually. I don't think that's even a release blocker though given that it's opt-in and limited surface area for now. We can see how much more API we get covered before 1.8.0 and make a call nearer the release.
  - I want to create new checklists in `contributing` for porting further blocking calls to use this. I think we probably need to do the first example on top of this first and document the process to enable further collaboration on porting.

### TODO Before Merge

Just a couple of things I'd like to do before merge:

 - [ ] End-to-end test this some more Ad-hoc and with internal testers.
 - [x] Re-run benchmarks against this branch to verify performance gains are still there.

### CODE REVIEW

File: agent/config/runtime_test.go [hunk 1]
```diff
@@ -4053,10 +4053,11 @@ func TestFullConfig(t *testing.T) {
 				},
 				"udp_answer_limit": 29909,
 				"use_cache": true,
-				"cache_max_age": "5m"` + entFullDNSJSONConfig + `
```
Comment (rboyer) [inline agent/config/runtime_test.go]: deliberate removal?

File: agent/consul/grpc_client.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			// Wrap the connection in a TLS client
+			tlsConn, err := wrapper(server.Datacenter, conn)
+			if err != nil {
+				conn.Close()
+				return nil, err
+			}
+			conn = tlsConn
+		}
+
+		_, err = conn.Write([]byte{pool.RPCGRPC})
```
Comment (rboyer) [inline agent/consul/grpc_client.go]: Caveat: I'm reading this commit-by-commit, so it may be addressed later.

Did you thread this through in such a way that wanfed + streaming would work?

File: agent/consul/grpc_resolver.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		// Shuffle the list of addresses using the last list given to the resolver.
+		resolver.addrLock.Lock()
+		addrs := resolver.lastAddrs
+		rand.Shuffle(len(addrs), func(i, j int) {
+			addrs[i], addrs[j] = addrs[j], addrs[i]
+		})
+		resolver.addrLock.Unlock()
+
+		// Pass the shuffled list to the resolver.
+		resolver.updateAddrs(addrs)
```
Comment (rboyer) [inline agent/consul/grpc_resolver.go]: Should this be in the locked section?

File: agent/consul/server.go [hunk 4]
```diff
@@ -861,6 +894,44 @@ func (s *Server) setupRPC() error {
...
+		conns: make(chan net.Conn),
+	}
+
+	// We don't need to pass tls.Config to the server since it's multiplexed
+	// behind the RPC listener, which already has TLS configured.
+	srv := grpc.NewServer(
+		grpc.StatsHandler(grpcStatsHandler),
+		grpc.StreamInterceptor(GRPCCountingStreamInterceptor),
+	)
+	//stream.RegisterConsulServer(srv, &ConsulGRPCAdapter{Health{s}})
```
Comment (rboyer) [inline agent/consul/server.go]: should this commented code be removed?

File: agent/agentpb/structgen/structgen.go [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+	f, err := os.OpenFile(name, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.Write(fCode)
+	if err != nil {
+		return err
+	}
+	f.Sync()
```
Comment (rboyer) [inline agent/agentpb/structgen/structgen.go]: This can return an error.

File: agent/agentpb/structs.proto [hunk 6]
```diff
@@ -0,0 +1,406 @@
+/**
+ * Structs.proto contains Protobug versions of the structs defined in
```
Comment (rboyer) [inline agent/agentpb/structs.proto]:
```suggestion
 * Structs.proto contains Protobuf versions of the structs defined in
```

File: agent/agentpb/subscribe.proto [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+import "gogoproto/gogo.proto";
+
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_getters_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+
+// Topic enumerates the supported streaming topics in Consul.
+enum Topic {
+    ServiceHealth = 0;
```
Comment (rboyer) [inline agent/agentpb/subscribe.proto]: When I used to do more protobuf RPC authoring, one thing that was considered a good idea was to avoid having the zero value for an enum be valid. So for example you could have the zero value for Topic be something like `Unknown = 0;` so that an omission of the topic due to bug or whatever doesn't silently operate as a `ServiceHealth` topic.

File: agent/agentpb/subscribe.proto [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+    // Key is a topic-specific identifier that restricts the scope of the
+    // subscription to only events pertaining to that identifier. For example
+    // for watching the instances of a single service, the service's name is
+    // specified as the key. An empty key indicates that all events in the topic
+    // are of interest.
+    string Key = 2;
+
+    // Token is the ACL token to authenticate the request. The token must have
+    // sufficient priviledges to read the requested information otherwise events
```
Comment (rboyer) [inline agent/agentpb/subscribe.proto]:
```suggestion
    // sufficient privileges to read the requested information otherwise events
```

File: agent/agentpb/subscribe.proto [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    // updates sent.
+    string Token = 3;
+
+    // Index is the raft index the subscriber has already observed up to. This
+    // is zero on an initial streaming call, but then can be provided by a
+    // client on subsequent reconnections such that the full snapshot doesn't
+    // need to be resent if the client is up to date.
+    uint64 Index = 4;
+
+    // Filter optionally specified a bexpr filter string to apply to the event
```
Comment (rboyer) [inline agent/agentpb/subscribe.proto]:
```suggestion
    // Filter optionally specifies a bexpr filter string to apply to the event
```

File: agent/agentpb/subscribe.proto [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    // Key is the logical identifier for the entity that was mutated.
+    string Key = 2;
+
+    // Index is the raft index at which the mutation took place. At the top
+    // level of a subscription there will always be at most one Event per index.
+    // If multiple events are published to the same topic in a single raft
+    // transaction then the batch of events will be encoded inside a single
+    // top-level event to ensure they are delivered atomically to clients.
+    uint64 Index = 3;
+
```
Comment (rboyer) [inline agent/agentpb/subscribe.proto]: Where did field ID `4` go?

File: agent/agentpb/subscribe.proto [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        ACLRoleUpdate ACLRole = 8;
+
+        // EndOfSnapshot indicates the event stream for the initial snapshot has
+        // ended. Subsequent Events delivered will be mutations to that result.
+        bool EndOfSnapshot = 9;
+
+        // ResumeStream indicates that the client is still up-to-date, so there
+        // will be no snapshot, and the rest of the stream will be individual
+        // update events. It disambiguates between "up to date, no snapshot" and
+        // "snapshot containes zero events".
```
Comment (rboyer) [inline agent/agentpb/subscribe.proto]:
```suggestion
        // "snapshot contains zero events".
```

File: agent/agentpb/subscribe.proto [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        bool ResumeStream = 10;
+
+        // ResetStream indicates the client should re-start the Subscribe call.
+        // It is used when the server can no longer correctly maintain the
+        // stream, for example because the ACL permisions for the token changed
```
Comment (rboyer) [inline agent/agentpb/subscribe.proto]:
```suggestion
        // stream, for example because the ACL permissions for the token changed
```

File: agent/consul/stream/event_buffer.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+// alowing efficient broadcast to many watchers without having to run multile
+// goroutines or deliver to O(N) separate channels.
+//
+// Because it's a linked list with atomically updated pointers, readers don't
+// have to take a lock and can consume at their own pace. but we also don't have
+// to have a fixed limit on the number of items which either means we don't have
+// to trade off buffer length config to balance using lots of memory wastefully
+// vs handling occasional slow readers.
+//
+// The buffer is used to deliver all messages broadcast toa topic for active
```
Comment (rboyer) [inline agent/consul/stream/event_buffer.go]:
```suggestion
// The buffer is used to deliver all messages broadcast to a topic for active
```

File: agent/consul/stream/event_buffer.go [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+// subscribers to consume, but it is also an effective way to both deliver and
+// optionally cache snapshots per topic and key. byt using an EventBuffer,
```
Comment (rboyer) [inline agent/consul/stream/event_buffer.go]:
```suggestion
// optionally cache snapshots per topic and key. by using an EventBuffer,
```

File: agent/connect/ca/provider_consul_test.go [hunk 15]
```diff
@@ -35,7 +35,7 @@ func (c *consulCAMockDelegate) ApplyCARequest(req *structs.CARequest) (interface
 
 		return true, nil
 	case structs.CAOpDeleteProviderState:
-		if err := c.state.CADeleteProviderState(req.ProviderState.ID); err != nil {
+		if err := c.state.CADeleteProviderState(idx+1, req.ProviderState.ID); err != nil {
```
Comment (rboyer) [inline agent/connect/ca/provider_consul_test.go]: Unrelated drive-by fix?

File: agent/consul/state/acl_events.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+// ACLEventsFromChanges returns all the ACL token, policy or role events that
+// should be emitted given a set of changes to the state store.
+func (s *Store) ACLEventsFromChanges(tx *txnWrapper, changes memdb.Changes) ([]agentpb.Event, error) {
+
+	// Don't allocate yet since in majority of update transactions no ACL token
+	// will be changed.
+	var events []agentpb.Event
+
+	getObj := func(change memdb.Change) interface{} {
```
Comment (rboyer) [inline agent/consul/state/acl_events.go]: These two local functions could be promoted to be first class regular functions instead of stack allocated closures.

File: agent/consul/state/catalog.go [hunk 17]
```diff
@@ -1486,18 +1486,10 @@ func (s *Store) ensureCheckTxn(tx *memdb.Txn, idx uint64, hc *structs.HealthChec
 			}
 		}
 	}
-	if modified {
-		// We update the modify index, ONLY if something has changed, thus
```
Comment (rboyer) [inline agent/consul/state/catalog.go]: Is there any value to retaining a portion of this large comment?

File: agent/consul/state/catalog_events.go [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+
+// _events.go files contain the streaming event methods relevant to specific
+// state store components.
+
+// ServiceHealthSnapshot is a stream.SnapFn that provides a streaming snapshot
+// of agentpb.Events that describe the current state of a service health query.
+func (s *Store) ServiceHealthSnapshot(req *agentpb.SubscribeRequest, buf *stream.EventBuffer) (uint64, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+	// TODO(banks): plumb entMeta through from SubscribeRequest
```
Comment (rboyer) [inline agent/consul/state/catalog_events.go]: suggestion: I've found it slightly nicer to use topical TODOs instead of personal ones, like `TODO(namespaces)` or `TODO(streaming)`.

File: agent/consul/state/memdb_wrapper.go [hunk 19]
```diff
@@ -0,0 +1,95 @@
+package state
+
+import (
+	"github.com/hashicorp/go-memdb"
+)
+
+// memDBWrapper is a thin shim over memdb.DB which forces all new tranactions to
```
Comment (rboyer) [inline agent/consul/state/memdb_wrapper.go]:
```suggestion
// memDBWrapper is a thin shim over memdb.DB which forces all new transactions to
```

File: agent/consul/state/memdb_wrapper.go [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+// Txn intercepts MemDB.Txn(). It allows read-only transactions to pass through
+// but fails write transactions as we now need to force callers to use a
+// slightly different API so that change capture can happen automatically. The
+// returned Txn object is wrapped with a no-op wrapper that just keeps all
+// transactions in our state store the same type. The wrapper only has
+// non-passthrough behavior for write transactions though.
+func (db *memDBWrapper) Txn(write bool) *txnWrapper {
+	if write {
+		panic("don't use db.Txn(true), use db.WriteTxn(idx uin64)")
```
Comment (rboyer) [inline agent/consul/state/memdb_wrapper.go]:
```suggestion
		panic("don't use db.Txn(true), use db.WriteTxn(idx uint64)")
```

File: agent/consul/subscribe_grpc_endpoint.go [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+
+	h.srv.logger.Trace("new subscription",
+		"topic", req.Topic.String(),
+		"key", req.Key,
+		"index", req.Index,
+		"stream_id", streamID,
+	)
+
+	var sentCount uint64
+	defer func() {
```
Comment (rboyer) [inline agent/consul/subscribe_grpc_endpoint.go]: why the extra `func() { }` wrapper?

File: agent/health_endpoint.go [hunk 22]
```diff
@@ -198,8 +198,13 @@ func (s *HTTPServer) healthServiceNodes(resp http.ResponseWriter, req *http.Requ
 	var out structs.IndexedCheckServiceNodes
 	defer setMeta(resp, &out.QueryMeta)
 
-	if args.QueryOptions.UseCache {
-		raw, m, err := s.agent.cache.Get(cachetype.HealthServicesName, &args)
+	// Use the streaming cachetype if enabled.
+	if args.QueryOptions.UseCache || (s.agent.config.EnableBackendStreaming && args.MinQueryIndex > 0) {
+		cType := cachetype.HealthServicesName
+		if s.agent.config.EnableBackendStreaming {
```
Comment (rboyer) [inline agent/health_endpoint.go]: Should this also include the `&& args.MinQueryIndex > 0` additional check?

**Comment by banks** [comment]
Benchmark re-run:

image

To recap what is being shown above:

 - 5000 simulated Consul agents (across 50 actual VMs) are all issuing one watch on a single service
 - That service has 2000 instances registered and one instance is being (de)registered every 10 second
 - This is closely modelled after a real user workload and is one of our motivating reasons to build streaming.

Notes on the results:
 - Mainly this looks much the same as previous streaming benchmarks which is the main thing here - this refactor has not lost any of that.
 - The new gRPC metrics are working still
 - We see server CPU down from 50-70% usage to about 20% during the initial snapshot delivery and almost none (in the noise of the background work on that server)for the steady state.
 - Network bandwidth goes from peaks of 2.5 Gbps to ~ 100Mbps during initial snapshot spike and then a few hundred kbps after that.
 - RAM goes from 1-4GB on servers down to ~300MB,
 - Number of goroutines is about the same and much more consistent thanks to long-lived streaming conns instead of constant "long" polls
 - The actual number of messages delivered (should be a sharp spike of 5000 every 10 seconds) is correct with streaming while blocking both smears it over time due to arrivals and seems to miss a few updates potentially due to clients being slow to process the large results each time. It's not missing data exactly, just updating more slowly. I assume this indicates streaming will improve the latency of updates being disseminated in this kind of case too based on that.
 - Note this workload and server size is specially chosen to be able to handle even the old blocking query load so raft times and server stability is not impacted for either workload.

File: agent/config/runtime_test.go (hunk 1 above)
Comment (banks) [reply agent/config/runtime_test.go:4056]: No, good spot. Fixed in next commit.

File: agent/consul/grpc_client.go (hunk 2 above)
Comment (banks) [reply agent/consul/grpc_client.go:115]: I _think_ so in that I added a new `pool.ALPN_RPCGRPC` too and handle that in the native TLS handler. Would be great to verify that though.

File: agent/consul/grpc_resolver.go (hunk 3 above)
Comment (banks) [reply agent/consul/grpc_resolver.go:85]: Great catch. I missed this in the original review.

I think it's outside intentionally because `updateAddrs` takes the same lock itself. Whether there is a chance of a race between releasing the lock we read with above and taking the lock to write below is another matter.

I hypothesise it could cause an issue it rebalance happens at just the same moment that Serf notifies of a change in servers - we could potentially read the old set and shuffle them, update resolver lastAddrs with the new set, then overwrite that with the shuffled old set again meaning we effectively "missed" that change in servers.

I'll change it so there is an `updateAddrsLocked` version of the method that we can use here to avoid the race.

Comment (banks) [reply agent/consul/grpc_resolver.go:85]: It's actually more complex since `updateAddrs` currently actually changes the underlying clientConn outside of the lock which could also produce races.

Comment (banks) [reply agent/consul/grpc_resolver.go:85]: Fixed in next commit!

File: agent/consul/server.go (hunk 4 above)
Comment (banks) [reply agent/consul/server.go]: It is in a later commit once the health endpoint actually exists again!

File: agent/agentpb/subscribe.proto (hunk 10 above)
Comment (banks) [reply agent/agentpb/subscribe.proto:92]: This has been refactored a few times. I could reset field numbers to be clean since it's new proto but Im not sure if it matters? Eventually fields will evolve. Do you think it's worth starting with them sequential?

File: agent/connect/ca/provider_consul_test.go (hunk 15 above)
Comment (banks) [reply agent/connect/ca/provider_consul_test.go:38]: No - we now require a raft index for every FSM operation the mutates state since `WriteTxn` requires one to capture it for event stuff.

So there were a few places where our state store methods performed writes but didn't take an index param because they didn't use it before and I had to change that.

File: agent/consul/state/acl_events.go (hunk 16 above)
Comment (banks) [reply agent/consul/state/acl_events.go:17]: Do you think they should be methods on `memdb.Change`? I could see that. They are kinda simple though my intention was just to try to keep the repetition/extra nesting out of the bodies of the switch here locally as it's already a lot to process. I'd tend to generalise them only once there was at least one other place where having a method in place of a single If statement was a real readability win? Changing memdb seems cleanest but also a bit of extra effort ;)

File: agent/consul/state/acl_events_test.go [hunk 23]
```diff
@@ -0,0 +1,342 @@
...
+		// 			return err
+		// 		}
+		// 		return nil
+		// 	},
+		// 	WantEvents: []agentpb.Event{
+		// 		// Should see a delete event
+		// 		testACLPolicyEvent(t, 100, 1, true),
+		// 	},
+		// 	WantErr: false,
+		// },
```
Comment (banks) [inline agent/consul/state/acl_events_test.go:232]: Err, these could be uncomented

File: agent/consul/subscribe_grpc_endpoint.go (hunk 21 above)
Comment (banks) [reply agent/consul/subscribe_grpc_endpoint.go:52]: I just love those curly braces. Fixed.

File: agent/health_endpoint.go (hunk 22 above)
Comment (banks) [reply agent/health_endpoint.go:204]: I think not since it's in the condition above - if we ever get here (say if the condition above changes one day) and streaming is enabled we should certainly use the streaming cache type.

File: agent/agentpb/subscribe.proto (hunk 12 above)
Comment (banks) [reply agent/agentpb/subscribe.proto:120]: btw. figured out how to enable my spell check for .proto file ;)

File: agent/consul/state/memdb_wrapper.go (hunk 19 above)
Comment (banks) [reply agent/consul/state/memdb_wrapper.go:7]: Turns out cSpell thinks "tranactions" is a valid word 🤷‍♂

**Comment by banks** [comment]
@rboyer thanks for your review - not an easy task on this one. Did you review the whole thing?

I think I addressed all of your comments either with a response or a change in the most recent commit.

I also rebased on master to fix a small conflict.

I realise this PR is too large to meaningfully review and the individual commits are cumbersome to work with in GH since the comments don't show in the normal files view and suggestions can't be committed from there etc.

I'll consider how to make this review process more approachable.

File: agent/consul/state/stream_publisher.go [hunk 24]
```diff
@@ -0,0 +1,439 @@
...
+		commitCh:        make(chan commitUpdate, 64),
+	}
+
+	// create a local handler table
+	for topic, handlers := range topicRegistry {
+		fnCopy := handlers.Snapshot
+		e.snapFns[topic] = func(req *agentpb.SubscribeRequest, buf *stream.EventBuffer) (uint64, error) {
+			return fnCopy(e.store, req, buf)
+		}
+	}
```
Comment (dnephin) [inline agent/consul/state/stream_publisher.go:88]: Is this done because the `topicRegistry` contains unbound functions, or is there another reason?

I was hoping to change this around and have `NewEventPublisher` accept a registry with bound functions.

File: agent/consul/state/stream_publisher.go [hunk 25]
```diff
@@ -0,0 +1,439 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+// secretHash returns a 256-bit Blake2 hash of the given string.
+func secretHash(token string) string {
+	hash, err := blake2b.New256(nil)
```
Comment (dnephin) [inline agent/consul/state/stream_publisher.go:301]: I'd like to add more to this docstring to explain why the token is being hashed. 

It is my understanding that Go is already going to hash the key used in a map. I don't think i understand why this is used to create map keys.

File: agent/consul/state/stream_publisher.go [hunk 26]
```diff
@@ -0,0 +1,439 @@
...
+			}
+			events = append(events, es...)
+		}
+	}
+	e.commitCh <- commitUpdate{
+		// Create a new transaction since it's going to be used from a different
+		// thread. Transactions aren't thread safe but it's OK to create it here
+		// since we won't try to use it in this thread and pass it straight to the
+		// handler which will own it exclusively.
+		tx:     e.store.db.Txn(false),
```
Comment (dnephin) [inline agent/consul/state/stream_publisher.go:161]: I don't understand the "since it's going to be used from a different thread" part. Why is it necessary to create this transaction here instead of from the goroutine? I suspect if I knew more about memdb it might make more sense to me.

Does this transaction need to be opened before the "active transaction" (the one that produced the changeset) is committed?

File: agent/consul/state/stream_publisher.go (hunk 24 above)
Comment (banks) [reply agent/consul/state/stream_publisher.go:88]: Yeah they are unbound as it makes it cleaner to use state store methods this way rather than stand-alone functions that need to be passed a ton of state. I don't feel very strongly, this is just the pattern we use elsewhere in the code base for RPC and HTTP handlers so I went with it.

Happy to change although I think the main thing to optimize for here is decoupling and easy maintenance and testing of the event handlers. Did you have a specific concern with this approach?

File: agent/consul/state/stream_publisher.go (hunk 25 above)
Comment (banks) [reply agent/consul/state/stream_publisher.go:301]: Yeah a comment would help here. The reason was that originally we were creating a map keyed by token and there was a concern that we were copying secrets around in memory unnecessarily which made them easier to accidentally leak in a debug message or similar.

I'm not super sure if that is relevant now since I _think_ a `string` key will end up being a pointer to the actual string literal that is already in memory in the state store anyway. It's also true that we store secrets in other structures currently - both in ACL caches and in the agent cache keys so on reflection I'm fine with removing it as I'm not sure it it a real security improvement within any realistic threat model. Arbitrary memory access on a server is already outside our threat model and the chances of us accidentally dumping this structure seem extremely low.

File: agent/consul/state/stream_publisher.go (hunk 26 above)
Comment (banks) [reply agent/consul/state/stream_publisher.go:161]: > Does this transaction need to be opened before the "active transaction" (the one that produced the changeset) is committed?

Yes it's necessary so that the event generating code is working with the exact same state of the store if it has to do any lookups that was true at the time the event was emitted. 

> Why is it necessary to create this transaction here instead of from the goroutine?
 
You basically figured it out with the question above - If we just passed execution to another goroutine here, returned from the FSM Apply and then let the publishing goroutine open a new transaction some time later, by the time it ran the state may have subsequently been changed and so the data it sees and publishes for an event would not be consistent with the state of the store at the time the event was actually emitted.

Originally, we forgot about thread safety entirely and passed the actual _same_ transaction across threads which caused lots of issues.

Another way to think about it is creating a read transaction like this is effectively taking a snapshot of the state store as it exists right now at the moment the event is published and passing it to be used by the publishing goroutine should it need to look up anything.

You picked up on the implicit part that should probably be added to the comment though. Something like:

```
// Create a new read-only transaction which becomes a snapshot of the state 
// at the time the Events were emitted that can be passed to the published in a 
// different goroutine. We can't pass the current active transaction as transaction 
// objects are not thread safe but we must open a new one here so that it is consistent 
// with the state at the point the events were emitted.
```

File: agent/consul/state/stream_publisher.go (hunk 24 above)
Comment (dnephin) [reply agent/consul/state/stream_publisher.go:88]: I've been thinking about the other places in the code base where we do this as well. [This diff](https://github.com/hashicorp/consul/compare/dnephin/http-api-separate-package) was an experiment I did to remove the pattern from the http handlers, and it appears to work.

> Happy to change although I think the main thing to optimize for here is decoupling and easy maintenance and testing of the event handlers.

I very much agree with this. I think the unbound method registry has a number of disadvantages, and no advantages I could find.

Declaring the registry statically (in init) makes it more difficult to test the generic parts that read the registry. If `New` accepts a map a test can pass it a set of stubs so the generic part can easily be tested independently of any of the specific implementations.

Requiring an unbound method forces us to implement every function on the same type. I think that leads to sprawling structs and packages. A registry of functions allows us to use functions from other packages.

We can keep the function signature small by passing in bound methods for the functions, which meets the goal of not passing around a ton of state.

File: agent/consul/state/stream_publisher.go (hunk 26 above)
Comment (dnephin) [reply agent/consul/state/stream_publisher.go:161]: Thanks! Sounds good

**Comment by banks** [comment]
@dnephin has been doing incredible work merging this in manageable chunks ready for 1.9 so it's time to close this one forever.
